---
title: "Rasmus"
output: html_document
date: "2024-08-22"
---

```{r setup, include=FALSE}
library(tidyverse)
library(ggplot2)
library(ggpubr)
library(ggpattern)
library(ggrepel)
library(stringr)
library(parafac4microbiome)
library(CMTFtoolbox)
library(scales)
```

```{r define colours}

# Old approach
# Needed colours: BMI groups (3) and phyla levels (6)
# colours = hue_pal()(9)
# #show_col(colours)
# BMI_cols = colours[c(1,4,7)]
# phylum_cols = colours[-c(1,4,7)]

# Rasmus's suggestion
colours = RColorBrewer:: brewer.pal(8, "Dark2")
BMI_cols = c("darkgreen","darkgoldenrod","darkred") #colours[1:3]
WHZ_cols = c("darkgreen", "darkgoldenrod", "dodgerblue4")
phylum_cols = colours#[-c(1:3)]
```

```{r check CV}
BMI1 = readRDS("./CV_small_BMI_0.25.RDS")
BMI2 = readRDS("./CV_small_BMI_0.5.RDS")
BMI3 = readRDS("./CV_small_BMI_0.75.RDS")
BMI4 = readRDS("./CV_small_BMI_0.9.RDS")
BMI5 = readRDS("./CV_small_BMI_0.95.RDS")
BMI6 = readRDS("./CV_small_BMI_1.RDS")
# investigateFMS(lapply(Jakobsen2025$Z$object, FUN=function(x){x@data}), Jakobsen2025$Z$modes, sharedMode=1, numCores=16, method="L-BFGS")
```

```{r plot the things}
plotCV = function(result){
  varExpData = result$varExp
  RMSEdata = result$RMSE
  a = varExpData %>%
    pivot_longer(-numComponents) %>%
    ggplot(ggplot2::aes(x = numComponents,y = value, col = as.factor(name))) +
    geom_path() +
    geom_point() +
    labs(x="Number of components", y="Variance explained (%)", color="Block") +
    scale_x_continuous(breaks=varExpData$numComponents) +
    theme(legend.position="na") +
    ylim(0,100)
  b = RMSEdata %>%
    ggplot(ggplot2::aes(x = numComponents, y = RMSE)) +
    geom_path() +
    geom_point() +
    labs(x = "Number of components", y = "RMSE") +
    scale_x_continuous(breaks = RMSEdata$numComponents)
  ggarrange(a,b)
}

plotCV(BMI1)
plotCV(BMI2)
plotCV(BMI3)
plotCV(BMI4)
plotCV(BMI5)
plotCV(BMI6)
```

```{r final prep}

Ybmi = Jakobsen2025$homogenizedSubjectMetadata$BMI
Ybmi_cnt = Ybmi - mean(Ybmi)
Ybmi_norm = Ybmi_cnt / norm(Ybmi_cnt,"2")
Ybmi_final = Ybmi_norm * sqrt(3)
```

```{r run models}
## ACMTF pi=1
ACMTF_models = readRDS("./ACMTF_model.RDS")
loss=do.call(rbind, lapply(ACMTF_models, FUN=function(x){x$f}))
bestIndex = which(loss == min(loss))
model_bmiX_2 = ACMTF_models[[bestIndex]]
# 
# # ACMTF-R pi=0.95
# ACMTFR_models = readRDS("./ACMTFR_model_bmi95.RDS")
# loss=do.call(rbind, lapply(ACMTFR_models, FUN=function(x){x$f}))
# bestIndex = which(loss == min(loss))
# model_bmi95 = ACMTFR_models[[bestIndex]]

# ACMTF-R pi=0.9

# ACMTF-R pi=0.75
ACMTFR_models = readRDS("./ACMTFR_model_bmi75.RDS")
loss75=do.call(rbind, lapply(ACMTFR_models, FUN=function(x){x$f}))
bestIndex = which(loss75 == min(loss75))
model_bmi75 = ACMTFR_models[[bestIndex]]

# ACMTF-R pi=0.50
ACMTFR_models = readRDS("./ACMTFR_model_bmi50.RDS")
loss50=do.call(rbind, lapply(ACMTFR_models, FUN=function(x){x$f}))
bestIndex = which(loss50 == min(loss50))
model_bmi50 = ACMTFR_models[[bestIndex]]

# ACMTF-R pi=0.25
ACMTFR_models = readRDS("./ACMTFR_model_bmi25.RDS")
loss25=do.call(rbind, lapply(ACMTFR_models, FUN=function(x){x$f}))
bestIndex = which(loss25 == min(loss25))
model_bmi25 = ACMTFR_models[[bestIndex]]
```

```{r plot the model}

plotModel = function(model, comp=1){
  a = Jakobsen2025$homogenizedSubjectMetadata %>% mutate(loading = model$Fac[[1]][,comp]) %>% arrange(BMI.group) %>% mutate(index=1:nrow(.)) %>% ggplot(aes(x=index,y=loading,fill=as.factor(BMI.group))) + geom_bar(stat="identity") + theme(legend.position = "na") + ylab(paste0("Component ", comp))

  b = abs(model$Fac[[8]][,comp]) %>% as_tibble() %>% mutate(Block = 1:nrow(.)) %>% ggplot(aes(x=Block,y=value)) + geom_bar(stat="identity") + ylab(expression(lambda))
  
  c = Jakobsen2025$faecal_microbiome_taxonomy %>% mutate(loading = model$Fac[[2]][,comp]) %>% arrange(V3) %>% mutate(index=1:nrow(.)) %>% ggplot(aes(x=index,y=loading,fill=as.factor(V3))) + geom_bar(stat="identity") + theme(legend.position = "na") + ylab(paste0("Component ", comp))
  
  d = model$Fac[[3]][,comp] %>% as_tibble() %>% mutate(index=1:nrow(.)) %>% ggplot(aes(x=index,y=value)) + geom_line() + geom_point() + ylab(paste0("Component ", comp))
  
  e = Jakobsen2025$milk_microbiome_taxonomy %>% mutate(loading = model$Fac[[4]][,comp]) %>% arrange(V3) %>% mutate(index=1:nrow(.)) %>% ggplot(aes(x=index,y=loading,fill=as.factor(V3))) + geom_bar(stat="identity") + theme(legend.position = "na") + ylab(paste0("Component ", comp))
  
  f = model$Fac[[5]][,comp] %>% as_tibble() %>% mutate(index=1:nrow(.)) %>% ggplot(aes(x=index,y=value)) + geom_line() + geom_point() + ylab(paste0("Component ", comp))
  
  g = Jakobsen2025$milk_metabolites %>% mutate(loading = model$Fac[[6]][,comp]) %>% arrange(X) %>% mutate(index=1:nrow(.)) %>% ggplot(aes(x=index,y=loading)) + geom_bar(stat="identity") + theme(legend.position = "na") + ylab(paste0("Component ", comp))
  
  h = model$Fac[[7]][,comp] %>% as_tibble() %>% mutate(index=1:nrow(.)) %>% ggplot(aes(x=index,y=value)) + geom_line() + geom_point() + ylab(paste0("Component ", comp))
  
  ggarrange(a,b,c,d,e,f,g,h, nrow=4, ncol=2)
}

plotModel(model_bmiX_2, 1)
plotModel(model_bmiX_2, 2)

plotModel(model_bmi25, 1)
plotModel(model_bmi50, 1)
plotModel(model_bmi75, 1)
```

```{r check associations of the subject mode}
# Import NPLS subject loadings
NPLS_faeces_mode1 = read.csv("./Data/NPLS_faeces_bmi_mode1.csv", header=FALSE) %>% as_tibble()
NPLS_milk_mode1 = read.csv("./Data/NPLS_milk_bmi_mode1.csv", header=FALSE) %>% as_tibble()
NPLS_milkMetab_mode1 = read.csv("./Data/NPLS_milkMetab_bmi_mode1.csv", header=FALSE) %>% as_tibble()
colnames(NPLS_faeces_mode1) = c("loading", "subject", "BMI.group")
colnames(NPLS_milk_mode1) = c("loading", "subject", "BMI.group")
colnames(NPLS_milkMetab_mode1) = c("loading", "subject", "BMI.group")

checkAssociations = function(loadings, comp=1){
  a = Jakobsen2025$homogenizedSubjectMetadata %>% mutate(loading = loadings) %>% ggplot(aes(x=loading,y=BMI)) + geom_point() + stat_cor()
  b = Jakobsen2025$homogenizedSubjectMetadata %>% mutate(loading = loadings) %>% ggplot(aes(x=as.factor(BMI.group),y=loading)) + geom_boxplot()
  c = Jakobsen2025$homogenizedSubjectMetadata %>% mutate(loading = loadings) %>% ggplot(aes(x=loading,y=whz.6m)) + geom_point() + stat_cor()
  d = Jakobsen2025$homogenizedSubjectMetadata %>% mutate(loading = loadings) %>% filter(C.section != "NA") %>% ggplot(aes(x=as.factor(C.section),y=loading)) + geom_boxplot() + stat_compare_means()
  e = Jakobsen2025$homogenizedSubjectMetadata %>% mutate(loading = loadings) %>% filter(Secretor != "NA") %>% ggplot(aes(x=as.factor(Secretor),y=loading)) + geom_boxplot() + stat_compare_means()
  f = Jakobsen2025$homogenizedSubjectMetadata %>% mutate(loading = loadings) %>% filter(Lewis != "NA") %>% ggplot(aes(x=as.factor(Lewis),y=loading)) + geom_boxplot() + stat_compare_means()
  ggarrange(a,b,c,d,e,f)
}

checkAssociations(model_bmiX_2$Fac[[1]][,1])
checkAssociations(model_bmiX_2$Fac[[1]][,2])
# checkAssociations(model_bmi95$Fac[[1]][,1])
checkAssociations(model_bmi75$Fac[[1]][,1])
checkAssociations(model_bmi50$Fac[[1]][,1])
checkAssociations(model_bmi25$Fac[[1]][,1])

# NPLS
checkAssociations(Jakobsen2025$homogenizedSubjectMetadata %>% left_join(NPLS_faeces_mode1 %>% mutate(subject = as.character(subject))) %>% select(loading) %>% pull())
checkAssociations(Jakobsen2025$homogenizedSubjectMetadata %>% left_join(NPLS_milk_mode1 %>% mutate(subject = as.character(subject))) %>% select(loading) %>% pull())
checkAssociations(Jakobsen2025$homogenizedSubjectMetadata %>% left_join(NPLS_milkMetab_mode1 %>% mutate(subject = as.character(subject))) %>% select(loading) %>% pull())
```

```{r plot the associations}
threshold = 0

# HM metabolome
df = Jakobsen2025$milk_metabolites %>% mutate(Component_1 = model_bmi95$Fac[[6]][,1])
colnames(df) = c("Metabolite", "CAS", "Component_1")

metaboliteCategories = read.csv("./Data/Metabolite_categories.txt", sep="\t") %>% as_tibble()
metaboliteCategories = metaboliteCategories %>% mutate(Metabolite = vctrs::vec_as_names(Metabolite, repair="universal_quiet"))
df = df %>% left_join(metaboliteCategories)

## Add mismatches by hand
df[df$Metabolite == "X2.Aminobutyrate","Class"] = "Amino acids and derivatives"
df[df$Metabolite == "X2.Fucosyllactose","Class"] = "Oligosaccharides"
df[df$Metabolite == "X2.Hydroxybutyrate","Class"] = "Amino acids and derivatives"
df[df$Metabolite == "X2.Oxoglutarate","Class"] = "Energy related"
df[df$Metabolite == "X3.Fucosyllactose","Class"] = "Oligosaccharides"
df[df$Metabolite == "X3SL.partial","Class"] = "Oligosaccharides"
df[df$Metabolite == "X6SL.partial","Class"] = "Oligosaccharides"
df[df$Metabolite == "Methionine","Class"] = "Amino acids and derivatives"

temp = df %>% arrange(Component_1) %>% filter(abs(Component_1) >= threshold) %>% mutate(index=1:nrow(.))
a = temp %>% ggplot(aes(x=Component_1,y=as.factor(index),fill=as.factor(Class))) + geom_bar_pattern(stat="identity", pattern_angle=45, pattern_density=0.05, pattern_spacing=0.02, pattern_key_scale_factor=0.6, col="black", pattern_fill="black") + ylab("Metabolite") + xlab("Component 1") + scale_y_discrete(labels=temp$Metabolite) + scale_fill_manual(name="Metabolite class", labels=c("Amino acids and derivatives", "Energy related", "Fatty acids and derivatives", "Food derived", "Microbially derived", "Oligosaccharides", "Pharmaceuticals", "Sugars"), values=phylum_cols) + theme(legend.position="none") + ylab("")


####

# Milk microbiome
df = Jakobsen2025$milk_microbiome_taxonomy %>% mutate(Component_1 = model_bmi95$Fac[[4]][,1]) %>% select(V3,V7,V8,Component_1)
colnames(df) = c("Phylum", "Genus", "Species", "Component_1")
temp = df %>% 
  arrange(Component_1) %>% 
  mutate(Genus = gsub("g__", "", Genus), Species = gsub("s__", "", Species)) %>% 
  mutate(Species = str_split_fixed(Species, "_", 2)[,2]) %>% 
  mutate(dplyr::across(.cols = Species, .fns = ~ dplyr::if_else(stringr::str_detect(.x, "sp."), "", .x))) %>%
  mutate(dplyr::across(.cols = Species, .fns = ~ dplyr::if_else(stringr::str_detect(.x, "organism"), "", .x))) %>%
  mutate(dplyr::across(.cols = Species, .fns = ~ dplyr::if_else(stringr::str_detect(.x, "bacterium"), "", .x)))

temp[temp$Species == "", "Species"] = "sp."
temp$name = paste0(temp$Genus, " ", temp$Species)
temp[temp$name == " sp.", "name"] = "Unclassified"
temp = temp %>% filter(abs(Component_1) >= threshold) %>% arrange(Component_1) %>% mutate(index=1:nrow(.))

b=temp %>% ggplot(aes(x=Component_1,y=as.factor(index),fill=as.factor(Phylum))) + geom_bar(stat="identity", col="black") + ylab("zOTU") + xlab("Component 1") + scale_y_discrete(labels=temp$name) + scale_fill_manual(name="Phylum", labels=c("Actinobacteriota", "Bacteroidota", "Firmicutes", "Proteobacteria"), values=phylum_cols[-3]) + theme(legend.position="none", axis.text.y = element_text(face="italic")) + ylab("")

# Faecal microbiome
df = Jakobsen2025$faecal_microbiome_taxonomy %>% mutate(Component_1 = model_bmi95$Fac[[2]][,1]) %>% select(V3,V7,V8,Component_1)
colnames(df) = c("Phylum", "Genus", "Species", "Component_1")
temp = df %>% 
  arrange(Component_1) %>% 
  mutate(Genus = gsub("g__", "", Genus), Species = gsub("s__", "", Species)) %>% 
  mutate(Species = str_split_fixed(Species, "_", 2)[,2]) %>% 
  mutate(dplyr::across(.cols = Species, .fns = ~ dplyr::if_else(stringr::str_detect(.x, "sp."), "", .x))) %>%
  mutate(dplyr::across(.cols = Species, .fns = ~ dplyr::if_else(stringr::str_detect(.x, "organism"), "", .x))) %>%
  mutate(dplyr::across(.cols = Species, .fns = ~ dplyr::if_else(stringr::str_detect(.x, "bacterium"), "", .x)))

temp[temp$Species == "", "Species"] = "sp."
temp$name = paste0(temp$Genus, " ", temp$Species)
temp[temp$name == " sp.", "name"] = "Unclassified"
temp = temp %>% filter(abs(Component_1) >= threshold) %>% arrange(Component_1) %>% mutate(index=1:nrow(.))

c=temp %>% ggplot(aes(x=Component_1,y=as.factor(index),fill=as.factor(Phylum))) + geom_bar(stat="identity", col="black") + ylab("zOTU") + xlab("Component 1") + scale_y_discrete(labels=temp$name) + scale_fill_manual(name="Phylum", labels=c("Actinobacteriota", "Bacteroidota", "Firmicutes", "Proteobacteria"), values=phylum_cols[-3]) + theme(legend.position="none", axis.text.y = element_text(face="italic")) + ylab("")

a
b
c
```
```{r loading loading plots}

# Faeces
NPLS_faeces = read.csv("./Data/NPLS_faeces_bmi_mode2.csv",header=FALSE) %>% as_tibble()
colnames(NPLS_faeces) = c("NPLS", "OTU", "Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species")
ACMTFR_faeces = cbind(-1*model_bmi95$Fac[[2]][,1], Jakobsen2025$faecal_microbiome_taxonomy) %>% as_tibble()
colnames(ACMTFR_faeces) = c("ACMTFR", "OTU", "Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species")

df = ACMTFR_faeces %>% left_join(NPLS_faeces)
  
temp = df %>% 
mutate(Genus = gsub("g__", "", Genus), Species = gsub("s__", "", Species)) %>% 
mutate(Species = str_split_fixed(Species, "_", 2)[,2]) %>% 
mutate(dplyr::across(.cols = Species, .fns = ~ dplyr::if_else(stringr::str_detect(.x, "sp."), "", .x))) %>%
mutate(dplyr::across(.cols = Species, .fns = ~ dplyr::if_else(stringr::str_detect(.x, "organism"), "", .x))) %>%
mutate(dplyr::across(.cols = Species, .fns = ~ dplyr::if_else(stringr::str_detect(.x, "bacterium"), "", .x)))

temp[temp$Species == "", "Species"] = "sp."
temp$name = paste0(temp$Genus, " ", temp$Species)
temp[temp$name == " sp.", "name"] = "Unclassified"

a = temp %>%
  ggplot(aes(x=ACMTFR,y=NPLS,col=as.factor(Phylum),label=name)) +
  geom_point() +
  # geom_text_repel(max.overlaps=10) +
  theme(legend.position="na")

# Milk
NPLS_milk = read.csv("./Data/NPLS_milk_bmi_mode2.csv",header=FALSE) %>% as_tibble()
colnames(NPLS_milk) = c("NPLS", "OTU", "Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species")
ACMTFR_milk = cbind(model_bmi95$Fac[[4]][,1], Jakobsen2025$milk_microbiome_taxonomy) %>% as_tibble()
colnames(ACMTFR_milk) = c("ACMTFR", "OTU", "Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species")

df = ACMTFR_milk %>% left_join(NPLS_milk)
  
temp = df %>% 
mutate(Genus = gsub("g__", "", Genus), Species = gsub("s__", "", Species)) %>% 
mutate(Species = str_split_fixed(Species, "_", 2)[,2]) %>% 
mutate(dplyr::across(.cols = Species, .fns = ~ dplyr::if_else(stringr::str_detect(.x, "sp."), "", .x))) %>%
mutate(dplyr::across(.cols = Species, .fns = ~ dplyr::if_else(stringr::str_detect(.x, "organism"), "", .x))) %>%
mutate(dplyr::across(.cols = Species, .fns = ~ dplyr::if_else(stringr::str_detect(.x, "bacterium"), "", .x)))

temp[temp$Species == "", "Species"] = "sp."
temp$name = paste0(temp$Genus, " ", temp$Species)
temp[temp$name == " sp.", "name"] = "Unclassified"

b = temp %>%
  ggplot(aes(x=ACMTFR,y=NPLS,col=as.factor(Phylum),label=name)) +
  geom_point() +
  # geom_text_repel(max.overlaps=10) +
  theme(legend.position="na")

# Milk metab
NPLS_milkMetab = read.csv("./Data/NPLS_milkMetab_bmi_mode2.csv", header=FALSE) %>% as_tibble()
colnames(NPLS_milkMetab) = c("NPLS", "Metabolite", "CAS")
ACMTFR_milkMetab = cbind(-1*model_bmi95$Fac[[6]][,1], Jakobsen2025$milk_metabolites) %>% as_tibble()
colnames(ACMTFR_milkMetab) = c("ACMTFR", "Metabolite", "CAS")

df = ACMTFR_milkMetab %>%
  left_join(NPLS_milkMetab, by="Metabolite") %>%
  left_join(metaboliteCategories)

## Add mismatches by hand
df[df$Metabolite == "X2.Aminobutyrate","Class"] = "Amino acids and derivatives"
df[df$Metabolite == "X2.Fucosyllactose","Class"] = "Oligosaccharides"
df[df$Metabolite == "X2.Hydroxybutyrate","Class"] = "Amino acids and derivatives"
df[df$Metabolite == "X2.Oxoglutarate","Class"] = "Energy related"
df[df$Metabolite == "X3.Fucosyllactose","Class"] = "Oligosaccharides"
df[df$Metabolite == "X3SL.partial","Class"] = "Oligosaccharides"
df[df$Metabolite == "X6SL.partial","Class"] = "Oligosaccharides"
df[df$Metabolite == "Methionine","Class"] = "Amino acids and derivatives"

c = df %>% 
  ggplot(aes(x=ACMTFR,y=NPLS,col=as.factor(Class),label=Metabolite)) +
  geom_point() +
  # geom_text_repel(max.overlaps=10) +
  theme(legend.position="na")

ggarrange(a,b,c, nrow=1)
```
```{r varExp plot}
# Load required packages
library(tidyverse)
library(ggpattern)
library(scales)  # for trans_new()

# 1. Define a custom transformation for a "broken" y axis.
#    Values <= breakpoint are unchanged; values > breakpoint are compressed.
#    Here we choose breakpoint = 15 and a factor = 5.67 so that 100 maps near 30.
broken_trans <- function(breakpoint = 15, factor = 5.67) {
    trans_new(
        name = paste0("broken_", breakpoint, "_", factor),
        transform = function(x) {
            ifelse(x <= breakpoint, x, breakpoint + (x - breakpoint) / factor)
        },
        inverse = function(x) {
            ifelse(x <= breakpoint, x, breakpoint + (x - breakpoint) * factor)
        },
        domain = c(0, Inf)
    )
}

# 2. Create the original 3x3 data matrix and assign column names.
df <- matrix(
    c(5.77, 5.31, 4.32,    # Block 1: values for NPLS, ACMTF, ACMTF-R
      12.85, 10.96, 5.90,    # Block 2
      1.08, 0.86, 1.15),    # Block 3
    nrow = 3, byrow = FALSE
)
colnames(df) <- c("NPLS", "ACMTF", "ACMTF-R")

# 3. Tidy the original data.
#    Create a column "orig_block" to track the original block number ("1", "2", or "3").
df_tidy <- as_tibble(df) %>%
    mutate(orig_block = as.character(1:3)) %>%  # original blocks
    pivot_longer(
        cols = -orig_block,
        names_to  = "method",
        values_to = "value"
    ) %>%
    # Here, the x-axis "block" is the original block.
    mutate(block = orig_block)

# 4. Add the extra block "Y" rows.
#    For NPLS: one extra row per block with the provided values.
df_extra_npls <- tibble(
    method     = "NPLS",
    block      = "Y",               # extra block
    value      = c(12.30, 11.22, 31.04),
    orig_block = c("1", "2", "3")    # so we know which NPLS facet it belongs to
)

#    For ACMTF-R: add one extra row.
df_extra_acmtfr <- tibble(
    method = "ACMTF-R",
    block  = "Y",
    value  = 98.86
)

#    For ACMTF: add one extra row with a value of 0.53.
df_extra_acmtf <- tibble(
    method = "ACMTF",
    block  = "Y",
    value  = 7.13
)

# 5. Combine all the data.
#    Specify the x-axis factor levels so that "Y" appears as the fourth level.
df_all <- bind_rows(df_tidy, df_extra_npls, df_extra_acmtfr, df_extra_acmtf) %>%
    mutate(block = factor(block, levels = c("1", "2", "3", "Y")))

# 6. Define custom x-axis labels.
#    Blocks 1–3 get descriptive labels; block "Y" is renamed to "Infant\n6-month\nWHZ".
block_labels <- c(
    "1" = "Infant\nfaecal\nmicrobiome",
    "2" = "Mother\nmilk\nmicrobiome",
    "3" = "Mother\nmilk\nmetabolome",
    "Y" = "Maternal\nppBMI"
)

# 7. Create a facet grouping variable.
#    For NPLS, each facet is labeled with the method and the corresponding original block label,
#    so that each NPLS facet only shows its own used blocks.
#    For ACMTF and ACMTF-R, the facet label remains simply the method name.
df_all <- df_all %>%
    mutate(facet_group = case_when(
        method == "NPLS" ~ paste("NPLS:", block_labels[orig_block]),
        TRUE           ~ method
    ))

# 8. Create the plot.
#    Here we use facet_grid with free x scales and free space so that the widths
#    of the facets are proportional to the number of x categories they display.
p <- ggplot(df_all, aes(x = block, y = value, fill = method)) +
  geom_col(position = position_dodge(width = 0.9), color = "black") +
  facet_grid(. ~ facet_group, scales = "free_x", space = "free_x") +
  scale_fill_manual(
    values = c("NPLS"    = "grey",
               "ACMTF"   = "grey",
               "ACMTF-R" = "grey")
  ) +
  scale_x_discrete(labels = block_labels, drop = TRUE) +
  xlab("") +
  ylab("Variance explained (%)") +
  # theme_bw(base_size = 12) +
  theme(legend.position = "none") +
  scale_y_continuous(
    limits = c(0, 100),
    trans  = broken_trans(breakpoint = 15, factor = 5.67),
    # Specify breaks in the original units.
    breaks = c(0, 5, 10, 15, 30, 50, 70, 90, 100),
    labels = c(0, 5, 10, 15, 30, 50, 70, 90, 100)
  )

p
``` 
