---
title: "Rasmus"
output: html_document
date: "2024-08-22"
---

```{r setup, include=FALSE}
library(tidyverse)
library(ggplot2)
library(ggpubr)
library(ggpattern)
library(ggrepel)
library(stringr)
library(parafac4microbiome)
library(NPLStoolbox)
library(CMTFtoolbox)
library(scales)
library(vegan)
```

```{r define colours}

# Old approach
# Needed colours: BMI groups (3) and phyla levels (6)
# colours = hue_pal()(9)
# #show_col(colours)
# BMI_cols = colours[c(1,4,7)]
# phylum_cols = colours[-c(1,4,7)]

# Rasmus's suggestion
colours = RColorBrewer:: brewer.pal(8, "Dark2")
BMI_cols = c("darkgreen","darkgoldenrod","darkred") #colours[1:3]
WHZ_cols = c("darkgreen", "darkgoldenrod", "dodgerblue4")
phylum_cols = colours
```

```{r process and homogenize data}
homogenizedSamples = intersect(intersect(NPLStoolbox::Jakobsen2025$faeces$mode1 %>% filter(!is.na(whz.6m)) %>% select(subject) %>% pull(), NPLStoolbox::Jakobsen2025$milkMicrobiome$mode1 %>% filter(!is.na(whz.6m)) %>% select(subject) %>% pull()), NPLStoolbox::Jakobsen2025$milkMetabolomics$mode1 %>% filter(!is.na(whz.6m)) %>% select(subject) %>% pull())

# Faeces
newFaeces = NPLStoolbox::Jakobsen2025$faeces

mask = newFaeces$mode1$subject %in% homogenizedSamples
newFaeces$data = newFaeces$data[mask,,]
newFaeces$mode1 = newFaeces$mode1[mask,]
processedFaeces = parafac4microbiome::processDataCube(newFaeces, sparsityThreshold = 0.75, centerMode=1, scaleMode=2)

# Milk
newMilk = NPLStoolbox::Jakobsen2025$milkMicrobiome

mask = newMilk$mode1$subject %in% homogenizedSamples
newMilk$data = newMilk$data[mask,,]
newMilk$mode1 = newMilk$mode1[mask,]
processedMilk = parafac4microbiome::processDataCube(newMilk, sparsityThreshold=0.85, centerMode=1, scaleMode=2)

# Milk metab
newMilkMetab = NPLStoolbox::Jakobsen2025$milkMetabolomics

mask = newMilkMetab$mode1$subject %in% homogenizedSamples
newMilkMetab$data = newMilkMetab$data[mask,,]
newMilkMetab$mode1 = newMilkMetab$mode1[mask,]
processedMilkMetab = parafac4microbiome::processDataCube(newMilkMetab, CLR=FALSE, centerMode=1, scaleMode=2)
```

```{r prepare y}
Y = processedFaeces$mode1$whz.6m
Ycnt = Y - mean(Y)
Ynorm = Ycnt / norm(Ycnt, "2")
```

```{r make NPLS models of the data}
# CV_faecal = NPLStoolbox::ncrossreg(processedFaeces$data, Ynorm)
# 
# a=CV_faecal$varExp %>% pivot_longer(-numComponents) %>% ggplot(aes(x=as.factor(numComponents),y=value,col=as.factor(name),group=as.factor(name))) + geom_line() + geom_point() + xlab("Number of components") + ylab("Variance explained (%)") + theme(legend.position="none")
# b=CV_faecal$RMSE %>% ggplot(aes(x=numComponents,y=RMSE)) + geom_line() + geom_point() + xlab("Number of components") + ylab("RMSECV")
# ggarrange(a,b)

NPLS_faecal = NPLStoolbox::triPLS1(processedFaeces$data, Ynorm, 1)

# CV_milk = NPLStoolbox::ncrossreg(processedMilk$data, Ynorm)
# 
# a=CV_milk$varExp %>% pivot_longer(-numComponents) %>% ggplot(aes(x=as.factor(numComponents),y=value,col=as.factor(name),group=as.factor(name))) + geom_line() + geom_point() + xlab("Number of components") + ylab("Variance explained (%)") + theme(legend.position="none")
# b=CV_milk$RMSE %>% ggplot(aes(x=numComponents,y=RMSE)) + geom_line() + geom_point() + xlab("Number of components") + ylab("RMSECV")
# ggarrange(a,b)

NPLS_milk = NPLStoolbox::triPLS1(processedMilk$data, Ynorm, 1)

# CV_milkMetab = NPLStoolbox::ncrossreg(processedMilkMetab$data, Ynorm)
# 
# a=CV_milkMetab$varExp %>% pivot_longer(-numComponents) %>% ggplot(aes(x=as.factor(numComponents),y=value,col=as.factor(name),group=as.factor(name))) + geom_line() + geom_point() + xlab("Number of components") + ylab("Variance explained (%)") + theme(legend.position="none")
# b=CV_milkMetab$RMSE %>% ggplot(aes(x=numComponents,y=RMSE)) + geom_line() + geom_point() + xlab("Number of components") + ylab("RMSECV")
# ggarrange(a,b)

NPLS_milkMetab = NPLStoolbox::triPLS1(processedMilkMetab$data, Ynorm, 2)
```

```{r create Z object for fitting ACMTF}
datasets = list(processedFaeces$data, processedMilk$data, processedMilkMetab$data)
modes = list(c(1,2,3), c(1,4,5), c(1,6,7))
Z = setupCMTFdata(datasets, modes, normalize=TRUE)

# model = CMTFtoolbox::acmtfr_opt(Z, Ynorm, numComponents=2, pi=1, nstart=1)
```

```{r check CV outcome}
# These files are very big and need to be cleared from RAM
# cv_outcome1 = readRDS("./CV_WHZ_ACMTF.RDS") # two components
# cv_outcome1$plots$overview
# rm(cv_outcome1)
# 
# cv_outcome95 = readRDS("./CV_small_WHZ_0.95.RDS") # two components
# cv_outcome95$plots$overview
# rm(cv_outcome95)
# 
# cv_outcome90 = readRDS("./CV_small_WHZ_0.9.RDS") # must be one due to degeneracy
# cv_outcome90$plots$overview
# rm(cv_outcome90)
# 
# cv_outcome89 = readRDS("./CV_small_WHZ_0.89.RDS") # must be one due to degeneracy
# cv_outcome89$plots$overview
# rm(cv_outcome89)
# 
# cv_outcome88 = readRDS("./CV_small_WHZ_0.88.RDS") # must be one due to degeneracy
# cv_outcome88$plots$overview
# rm(cv_outcome88)
# 
# cv_outcome87 = readRDS("./CV_small_WHZ_0.87.RDS") # must be one due to degeneracy
# cv_outcome87$plots$overview
# rm(cv_outcome87)
# 
# cv_outcome86 = readRDS("./CV_small_WHZ_0.86.RDS") # must be one due to degeneracy
# cv_outcome86$plots$overview
# rm(cv_outcome86)
# 
# cv_outcome85 = readRDS("./CV_small_WHZ_0.85.RDS") # must be one due to degeneracy
# cv_outcome85$plots$overview
# rm(cv_outcome85)
# 
# cv_outcome80 = readRDS("./CV_small_WHZ_0.8.RDS") # must be one due to degeneracy, overfit
# cv_outcome80$plots$overview
# rm(cv_outcome80)
# 
# cv_outcome75 = readRDS("./CV_small_WHZ_0.75.RDS") # must be one due to degeneracy, overfit
# cv_outcome75$plots$overview
# rm(cv_outcome75)
# 
# cv_outcome50 = readRDS("./CV_small_WHZ_0.5.RDS") # must be one due to degeneracy, overfit
# cv_outcome50$plots$overview
# rm(cv_outcome50)
# 
# cv_outcome25 = readRDS("./CV_small_WHZ_0.25.RDS") # nonsensical
# cv_outcome25$plots$overview
# rm(cv_outcome25)
```

```{r find acmtf model}
ACMTFR_models = readRDS("./ACMTFR_model_whz1.RDS")

iter = do.call(rbind, lapply(ACMTFR_models, FUN=function(x){x$iter}))
loss = do.call(rbind, lapply(ACMTFR_models, FUN=function(x){x$f}))
varExpX = do.call(rbind, lapply(ACMTFR_models, FUN=function(x){x$varExp}))
varExpY = do.call(rbind, lapply(ACMTFR_models, FUN=function(x){x$varExpY}))

df = cbind(loss, varExpX, varExpY) %>% as_tibble()
colnames(df) = c("Loss", "varExpX1", "varExpX2", "varExpX3", "varExpY")

df %>% mutate(index=1:nrow(.)) %>% pivot_longer(-index) %>% mutate(name=factor(name, levels=c("varExpX1", "varExpX2", "varExpX3", "Loss", "varExpY"))) %>% ggplot(aes(x=index,y=value)) + facet_wrap(~name,scale="free_y") + geom_bar(stat="identity") + scale_y_log10()

df %>% mutate(index=1:nrow(.)) %>% arrange(Loss)

bestIndex = 24
model_ACMTF = ACMTFR_models[[bestIndex]]
```
```{r find best model pi=0.95}
ACMTFR_models = readRDS("./ACMTFR_model_whz95.RDS")

iter = do.call(rbind, lapply(ACMTFR_models, FUN=function(x){x$iter}))
loss = do.call(rbind, lapply(ACMTFR_models, FUN=function(x){x$f}))
varExpX = do.call(rbind, lapply(ACMTFR_models, FUN=function(x){x$varExp}))
varExpY = do.call(rbind, lapply(ACMTFR_models, FUN=function(x){x$varExpY}))
# cor1_2 = do.call(rbind, lapply(ACMTFR_models, FUN=function(x){cor(x$Fac[[1]][,1],x$Fac[[1]][,2])}))

df = cbind(loss, varExpX, varExpY) %>% as_tibble()
colnames(df) = c("Loss", "varExpX1", "varExpX2", "varExpX3", "varExpY")

df %>% mutate(index=1:nrow(.)) %>% pivot_longer(-index) %>% ggplot(aes(x=index,y=value)) + facet_wrap(~name,scale="free_y") + geom_bar(stat="identity")

df %>% mutate(index=1:nrow(.)) %>% arrange(Loss)

bestIndex = 10
model_whz95 = ACMTFR_models[[bestIndex]]
```

```{r find best model pi=0.90}
ACMTFR_models = readRDS("./ACMTFR_model_whz90.RDS")

iter = do.call(rbind, lapply(ACMTFR_models, FUN=function(x){x$iter}))
loss = do.call(rbind, lapply(ACMTFR_models, FUN=function(x){x$f}))
varExpX = do.call(rbind, lapply(ACMTFR_models, FUN=function(x){x$varExp}))
varExpY = do.call(rbind, lapply(ACMTFR_models, FUN=function(x){x$varExpY}))
# cor1_2 = do.call(rbind, lapply(ACMTFR_models, FUN=function(x){cor(x$Fac[[1]][,1],x$Fac[[1]][,2])}))

df = cbind(loss, varExpX, varExpY) %>% as_tibble()
colnames(df) = c("Loss", "varExpX1", "varExpX2", "varExpX3", "varExpY")

df %>% mutate(index=1:nrow(.)) %>% pivot_longer(-index) %>% ggplot(aes(x=index,y=value)) + facet_wrap(~name,scale="free_y") + geom_bar(stat="identity")

df %>% mutate(index=1:nrow(.)) %>% arrange(Loss)

bestIndex = 91
model_whz90 = ACMTFR_models[[bestIndex]]
```

```{r find best model pi=0.85}
ACMTFR_models = readRDS("./ACMTFR_model_whz85.RDS")

iter = do.call(rbind, lapply(ACMTFR_models, FUN=function(x){x$iter}))
loss = do.call(rbind, lapply(ACMTFR_models, FUN=function(x){x$f}))
varExpX = do.call(rbind, lapply(ACMTFR_models, FUN=function(x){x$varExp}))
varExpY = do.call(rbind, lapply(ACMTFR_models, FUN=function(x){x$varExpY}))
# cor1_2 = do.call(rbind, lapply(ACMTFR_models, FUN=function(x){cor(x$Fac[[1]][,1],x$Fac[[1]][,2])}))

df = cbind(loss, varExpX, varExpY) %>% as_tibble()
colnames(df) = c("Loss", "varExpX1", "varExpX2", "varExpX3", "varExpY")

df %>% mutate(index=1:nrow(.)) %>% pivot_longer(-index) %>% ggplot(aes(x=index,y=value)) + facet_wrap(~name,scale="free_y") + geom_bar(stat="identity")

df %>% mutate(index=1:nrow(.)) %>% arrange(Loss)

bestIndex = 88
model_whz85 = ACMTFR_models[[bestIndex]]
```

```{r find best model pi=0.80}
ACMTFR_models = readRDS("./ACMTFR_model_whz80.RDS")

iter = do.call(rbind, lapply(ACMTFR_models, FUN=function(x){x$iter}))
loss = do.call(rbind, lapply(ACMTFR_models, FUN=function(x){x$f}))
varExpX = do.call(rbind, lapply(ACMTFR_models, FUN=function(x){x$varExp}))
varExpY = do.call(rbind, lapply(ACMTFR_models, FUN=function(x){x$varExpY}))
# cor1_2 = do.call(rbind, lapply(ACMTFR_models, FUN=function(x){cor(x$Fac[[1]][,1],x$Fac[[1]][,2])}))

df = cbind(loss, varExpX, varExpY) %>% as_tibble()
colnames(df) = c("Loss", "varExpX1", "varExpX2", "varExpX3", "varExpY")

df %>% mutate(index=1:nrow(.)) %>% pivot_longer(-index) %>% ggplot(aes(x=index,y=value)) + facet_wrap(~name,scale="free_y") + geom_bar(stat="identity")

df %>% mutate(index=1:nrow(.)) %>% arrange(Loss)

bestIndex = 26
model_whz80 = ACMTFR_models[[bestIndex]]
```

```{r find best model pi=0.75}
ACMTFR_models = readRDS("./ACMTFR_model_whz75.RDS")

iter = do.call(rbind, lapply(ACMTFR_models, FUN=function(x){x$iter}))
loss = do.call(rbind, lapply(ACMTFR_models, FUN=function(x){x$f}))
varExpX = do.call(rbind, lapply(ACMTFR_models, FUN=function(x){x$varExp}))
varExpY = do.call(rbind, lapply(ACMTFR_models, FUN=function(x){x$varExpY}))
# cor1_2 = do.call(rbind, lapply(ACMTFR_models, FUN=function(x){cor(x$Fac[[1]][,1],x$Fac[[1]][,2])}))

df = cbind(loss, varExpX, varExpY) %>% as_tibble()
colnames(df) = c("Loss", "varExpX1", "varExpX2", "varExpX3", "varExpY")

df %>% mutate(index=1:nrow(.)) %>% pivot_longer(-index) %>% ggplot(aes(x=index,y=value)) + facet_wrap(~name,scale="free_y") + geom_bar(stat="identity")

df %>% mutate(index=1:nrow(.)) %>% arrange(Loss)

bestIndex = 79
model_whz75 = ACMTFR_models[[bestIndex]]
```

```{r find best model pi=0.50}
ACMTFR_models = readRDS("./ACMTFR_model_whz50.RDS")

iter = do.call(rbind, lapply(ACMTFR_models, FUN=function(x){x$iter}))
loss = do.call(rbind, lapply(ACMTFR_models, FUN=function(x){x$f}))
varExpX = do.call(rbind, lapply(ACMTFR_models, FUN=function(x){x$varExp}))
varExpY = do.call(rbind, lapply(ACMTFR_models, FUN=function(x){x$varExpY}))
# cor1_2 = do.call(rbind, lapply(ACMTFR_models, FUN=function(x){cor(x$Fac[[1]][,1],x$Fac[[1]][,2])}))

df = cbind(loss, varExpX, varExpY) %>% as_tibble()
colnames(df) = c("Loss", "varExpX1", "varExpX2", "varExpX3", "varExpY")

df %>% mutate(index=1:nrow(.)) %>% pivot_longer(-index) %>% ggplot(aes(x=index,y=value)) + facet_wrap(~name,scale="free_y") + geom_bar(stat="identity")

df %>% mutate(index=1:nrow(.)) %>% arrange(Loss)

bestIndex = 78
model_whz50 = ACMTFR_models[[bestIndex]]
```

```{r plot the model}

plotModel = function(model, comp=1){
  a = processedFaeces$mode1 %>%
    mutate(loading = model$Fac[[1]][,comp]) %>%
    mutate(BMI.group=factor(BMI.group,levels=c("NW","OW","OB"))) %>%
    arrange(BMI.group) %>%
    mutate(index=1:nrow(.)) %>%
    ggplot(aes(x=index,y=loading,fill=as.factor(BMI.group))) +
    geom_bar(stat="identity") +
    scale_fill_manual(name="BMI group", labels=c("Normal", "Overweight", "Obese"), values=BMI_cols) +
    theme(legend.position = "na") +
    ylab(paste0("Component ", comp)) +
    xlab("Subject index") +
    ggtitle("Subject mode")

    b = abs(model$Fac[[8]][,comp]) %>%
    as_tibble() %>%
    mutate(Block = paste0("X", 1:nrow(.))) %>%
    ggplot(aes(x=Block,y=value)) +
    geom_bar(stat="identity") +
    ylab(expression(lambda)) +
    ggtitle("Lambda")
  
  c = processedFaeces$mode2 %>%
    mutate(loading = model$Fac[[2]][,comp]) %>%
    arrange(V3) %>%
    mutate(index=1:nrow(.)) %>%
    ggplot(aes(x=index,y=loading,fill=as.factor(V3))) +
    geom_bar(stat="identity") +
    scale_fill_manual(name="Phylum", labels=c("Actinobacteriota", "Bacteroidota", "Firmicutes", "Proteobacteria", "Verrucomicrobiota"), values=phylum_cols[c(1,2,4,5,6)]) +
    theme(legend.position = "na") +
    ylab(paste0("Component ", comp)) +
    xlab("Feature index") +
    ggtitle("Faecal microbiota")
  
  d = model$Fac[[3]][,comp] %>%
    as_tibble() %>%
    mutate(index=1:nrow(.)) %>%
    ggplot(aes(x=index,y=value)) +
    geom_line() + geom_point() +
    ylab(paste0("Component ", comp)) +
    scale_x_continuous(name="Time point [days]", breaks=c(1:3), labels=c(30,60,90)) +
    ggtitle("Faecal microbiota")
  
  e = processedMilk$mode2 %>%
    mutate(loading = model$Fac[[4]][,comp]) %>%
    arrange(V3) %>%
    mutate(index=1:nrow(.)) %>%
    ggplot(aes(x=index,y=loading,fill=as.factor(V3))) +
    geom_bar(stat="identity") +
    scale_fill_manual(name="Phylum", labels=c("Actinobacteriota", "Bacteroidota", "Cyanobacteria", "Firmicutes", "Proteobacteria", "Verrucomicrobiota"), values=phylum_cols[1:6]) +
    theme(legend.position = "na") +
    ylab(paste0("Component ", comp)) +
    xlab("Feature index") +
    ggtitle("HM microbiota")
  
  f = model$Fac[[5]][,comp] %>%
    as_tibble() %>%
    mutate(index=1:nrow(.)) %>%
    ggplot(aes(x=index,y=value)) +
    geom_line() +
    geom_point() +
    ylab(paste0("Component ", comp)) +
    scale_x_continuous(name="Time point [days]", breaks=c(1:4), labels=c(3,30,60,90)) +
    ggtitle("HM microbiota")
  
  g = processedMilkMetab$mode2 %>%
    mutate(loading = model$Fac[[6]][,comp]) %>%
    arrange(Class,Metabolite) %>%
    mutate(index=1:nrow(.)) %>%
    ggplot(aes(x=index,y=loading,fill=as.factor(Class))) +
    geom_bar(stat="identity") +
    scale_fill_manual(name="Metabolite class", labels=c("Amino acids and derivatives", "Energy related", "Fatty acids and derivatives", "Food derived", "Microbially derived", "Oligosaccharides", "Pharmaceuticals", "Sugars"), values=phylum_cols) +
    theme(legend.position = "na") +
    ylab(paste0("Component ", comp)) +
    xlab("Feature index") +
    ggtitle("HM metabolomics")
  
  h = model$Fac[[7]][,comp] %>%
    as_tibble() %>%
    mutate(index=1:nrow(.)) %>%
    ggplot(aes(x=index,y=value)) +
    geom_line() +
    geom_point() +
    ylab(paste0("Component ", comp)) +
    scale_x_continuous(name="Time point [days]", breaks=c(1:4), labels=c(3,30,60,90)) +
    ggtitle("HM metabolomics")
  
  ggarrange(a,b,c,d,e,f,g,h, nrow=4, ncol=2)
}

plotModel(model_ACMTF, 1)

plotModel(model_whz95, 1)

plotModel(model_whz90, 1)

plotModel(model_whz85, 1)

plotModel(model_whz80, 1)
```

```{r plot the NPLS models}
# Faecal
a=processedFaeces$mode1 %>% mutate(Component_1=NPLS_faecal$Fac[[1]][,1]) %>% arrange(BMI.group,subject) %>% mutate(index=1:nrow(.)) %>% ggplot(aes(x=index,y=Component_1,fill=as.factor(BMI.group))) + geom_bar(stat="identity") + scale_fill_manual(name="BMI group", labels=c("Normal", "Overweight", "Obese"), values=BMI_cols) + theme(legend.position="none") + xlab("Subject index") + ylab("Component 1")

b=processedFaeces$mode2 %>% mutate(Component_1=NPLS_faecal$Fac[[2]][,1]) %>% arrange(V3,V4,V5,V6,V7,V8,V1) %>% mutate(index=1:nrow(.)) %>% ggplot(aes(x=index,y=Component_1,fill=as.factor(V3))) + geom_bar(stat="identity") + scale_fill_manual(name="Phylum", labels=c("Actinobacteriota", "Bacteroidota", "Firmicutes", "Proteobacteria", "Verrucomicrobiota"), values=phylum_cols[c(1,2,4,5,6)]) + theme(legend.position="none") + xlab("Feature index") + ylab("Component 1")

c=processedFaeces$mode3 %>% mutate(Component_1=NPLS_faecal$Fac[[3]][,1]) %>% ggplot(aes(x=timepoint,y=Component_1)) + geom_line() + geom_point()+ xlab("Time point [days]") + ylab("Component 1")

# Milk
d=processedMilk$mode1 %>% mutate(Component_1=NPLS_milk$Fac[[1]][,1]) %>% arrange(BMI.group,subject) %>% mutate(index=1:nrow(.)) %>% ggplot(aes(x=index,y=Component_1,fill=as.factor(BMI.group))) + geom_bar(stat="identity") + scale_fill_manual(name="BMI group", labels=c("Normal", "Overweight", "Obese"), values=BMI_cols) + theme(legend.position="none") + xlab("Subject index") + ylab("Component 1")

e=processedMilk$mode2 %>% mutate(Component_1=NPLS_milk$Fac[[2]][,1]) %>% arrange(V3,V4,V5,V6,V7,V8,V1) %>% mutate(index=1:nrow(.)) %>% ggplot(aes(x=index,y=Component_1,fill=as.factor(V3))) + geom_bar(stat="identity") + scale_fill_manual(name="Phylum", labels=c("Actinobacteriota", "Bacteroidota", "Cyanobacteria", "Firmicutes", "Proteobacteria", "Verrucomicrobiota"), values=phylum_cols[1:6]) + theme(legend.position="none") + xlab("Feature index") + ylab("Component 1")

f=processedMilk$mode3 %>% mutate(Component_1=NPLS_milk$Fac[[3]][,1]) %>% ggplot(aes(x=timepoint,y=Component_1)) + geom_line() + geom_point() + xlab("Time point [days]") + ylab("Component 1")

# MilkMetab
g=processedMilkMetab$mode1 %>% mutate(Component_1=NPLS_milkMetab$Fac[[1]][,1]) %>% arrange(BMI.group,subject) %>% mutate(index=1:nrow(.)) %>% ggplot(aes(x=index,y=Component_1,fill=as.factor(BMI.group))) + geom_bar(stat="identity") + scale_fill_manual(name="BMI group", labels=c("Normal", "Overweight", "Obese"), values=BMI_cols) +theme(legend.position="none") + xlab("Subject index") + ylab("Component 1")

h=processedMilkMetab$mode2 %>% mutate(Component_1=NPLS_milkMetab$Fac[[2]][,1]) %>% arrange(Class,Metabolite) %>% mutate(index=1:nrow(.)) %>% ggplot(aes(x=index,y=Component_1,fill=as.factor(Class))) + geom_bar(stat="identity") + scale_fill_manual(name="Metabolite class", labels=c("Amino acids and derivatives", "Energy related", "Fatty acids and derivatives", "Food derived", "Microbially derived", "Oligosaccharides", "Pharmaceuticals", "Sugars"), values=phylum_cols) + theme(legend.position="none") + xlab("Feature index") + ylab("Component 1")

i=processedMilkMetab$mode3 %>% mutate(Component_1=NPLS_milkMetab$Fac[[3]][,1]) %>% ggplot(aes(x=timepoint,y=Component_1)) + geom_line() + geom_point() + xlab("Time point [days]") + ylab("Component 1")

ggarrange(a,b,c,d,e,f,g,h,i, nrow=3, ncol=3)
```

```{r test overall associations using PERMANOVA}
set.seed(123)

testMetadata = function(model, metadata){
  transformedSubjectLoadings = transformPARAFACloadings(model$Fac, 1)
  
  result = adonis2(transformedSubjectLoadings ~ BMI + whz.6m + C.section + Secretor + Lewis, data=metadata, method="euclidean", by="margin", permutations=9999, na.action=na.omit)
  
  return(result)
}

testMetadata(NPLS_faecal, newFaeces$mode1)
testMetadata(NPLS_milk, newFaeces$mode1)
testMetadata(NPLS_milkMetab, newFaeces$mode1)

testMetadata(model_whz50, newFaeces$mode1)
testMetadata(model_whz75, newFaeces$mode1)
testMetadata(model_whz80, newFaeces$mode1)
testMetadata(model_whz85, newFaeces$mode1)
testMetadata(model_whz90, newFaeces$mode1)
testMetadata(model_whz95, newFaeces$mode1)
testMetadata(model_ACMTF, newFaeces$mode1)

uncorrectedP = matrix(c(0.1113,0.0106,0.2240,0.5513,0.0989,0.0090,0.0002,0.2506,0.6816,0.8875,0.0165,0.0030,0.7277,0.0001,0.0062,0.7541,0.0001,0.7495,0.0227,0.5829,0.8193,0.0001,0.5866,0.0482,0.4912,0.4453,0.0001,0.4736,0.1216,0.3850,0.0178,0.0001,0.3016,0.6917,0.0708,0.0004,0.8437,0.2919,0.1341,0.0596,0.0007,0.5690,0.2213,0.0001,0.0602,0.0005,0.8317,0.2078,0.0001,0.0685), nrow=5, ncol=10)
apply(uncorrectedP, 2, FUN=function(x){p.adjust(x,"BH")}) # BH per dataset

temp = newFaeces$mode1 %>% mutate(Component_1 = NPLS_milkMetab$Fac[[1]][,1], Component_2 = NPLS_milkMetab$Fac[[1]][,2])
summary(lm(BMI ~ Component_1 + Component_2, data=temp))
summary(lm(whz.6m ~ Component_1 + Component_2, data=temp))
summary(lm(Secretor ~ Component_1 + Component_2, data=temp))
summary(lm(Lewis ~ Component_1 + Component_2, data=temp))

temp = newFaeces$mode1 %>% mutate(Component_1 = model_whz95$Fac[[1]][,1], Component_2 = model_whz95$Fac[[1]][,2])
summary(lm(BMI ~ Component_1 + Component_2, data=temp))
summary(lm(Secretor ~ Component_1 + Component_2, data=temp))

temp = newFaeces$mode1 %>% mutate(Component_1 = model_ACMTF$Fac[[1]][,1], Component_2 = model_ACMTF$Fac[[1]][,2])
summary(lm(BMI ~ Component_1 + Component_2, data=temp))
summary(lm(Secretor ~ Component_1 + Component_2, data=temp))
```

```{r test associations per component}
# Import NPLS subject loadings
# NPLS_faeces_mode1 = read.csv("./Data/NPLS_faeces_bmi_mode1.csv", header=FALSE) %>% as_tibble()
# NPLS_milk_mode1 = read.csv("./Data/NPLS_milk_bmi_mode1.csv", header=FALSE) %>% as_tibble()
# NPLS_milkMetab_mode1 = read.csv("./Data/NPLS_milkMetab_bmi_mode1.csv", header=FALSE) %>% as_tibble()
# colnames(NPLS_faeces_mode1) = c("loading", "subject", "BMI.group")
# colnames(NPLS_milk_mode1) = c("loading", "subject", "BMI.group")
# colnames(NPLS_milkMetab_mode1) = c("loading", "subject", "BMI.group")
# 
# NPLS_faeces_bmi = list()
# NPLS_faeces_bmi$Fac = list()
# NPLS_faeces_bmi$Fac[[1]] = processedFaeces$mode1 %>% left_join(NPLS_faeces_mode1 %>% mutate(subject = as.character(subject))) %>% select(loading) %>% as.matrix()
# 
# NPLS_milk_bmi = list()
# NPLS_milk_bmi$Fac = list()
# NPLS_milk_bmi$Fac[[1]] = processedMilk$mode1 %>% left_join(NPLS_milk_mode1 %>% mutate(subject = as.character(subject))) %>% select(loading) %>% as.matrix()
# 
# NPLS_milkMetab_bmi = list()
# NPLS_milkMetab_bmi$Fac = list()
# NPLS_milkMetab_bmi$Fac[[1]] = processedMilkMetab$mode1 %>% left_join(NPLS_milkMetab_mode1 %>% mutate(subject = as.character(subject))) %>% select(loading) %>% as.matrix()

testAssociations2 = function(model, metadata){
  transformedSubjectLoadings = parafac4microbiome::transformPARAFACloadings(model$Fac, 1)
  
  result = adonis2(transformedSubjectLoadings ~ BMI + whz.6m + C.section + Secretor + Lewis, data=metadata, method="euclidean", by="margin", permutations=9999, na.action=na.omit)
  return(result)
}

testAssociationsPerComponent = function(model){
  numComponents = ncol(model$Fac[[1]])
  transformedSubjectLoadings = parafac4microbiome::transformPARAFACloadings(model$Fac, 1)
  
  result = matrix(NA, nrow=numComponents, ncol=5)
  
  for(i in 1:numComponents){
    result[i,1] = cor.test(transformedSubjectLoadings[,i], processedFaeces$mode1$BMI, method="spearman")$p.value
    result[i,2] = cor.test(transformedSubjectLoadings[,i], processedFaeces$mode1$whz.6m, method="spearman")$p.value
    
    mask = !is.na(processedFaeces$mode1$C.section)
    temp = transformedSubjectLoadings[mask,i]
    tempMeta = processedFaeces$mode1[mask,]
    mask1 = tempMeta$C.section == 1
    mask2 = tempMeta$C.section == 2
    
    result[i,3] = wilcox.test(temp[mask1], temp[mask2])$p.value
    
    mask1 = processedFaeces$mode1$Secretor == 0
    mask2 = processedFaeces$mode1$Secretor == 1
    result[i,4] = wilcox.test(transformedSubjectLoadings[mask1,i], transformedSubjectLoadings[mask2,i])$p.value
    
    mask1 = processedFaeces$mode1$Lewis == 0
    mask2 = processedFaeces$mode1$Lewis == 1
    result[i,5] = wilcox.test(transformedSubjectLoadings[mask1,i], transformedSubjectLoadings[mask2,i])$p.value
  }
  
  return(result)
}

# testAssociations2(NPLS_faecal, processedFaeces$mode1)
# testAssociations2(NPLS_milk, processedFaeces$mode1)
# testAssociations2(NPLS_milkMetab, processedFaeces$mode1)
# testAssociations2(model_whz80, processedFaeces$mode1)
# testAssociations2(model_whz85, processedFaeces$mode1)
# testAssociations2(model_whz90, processedFaeces$mode1)
# testAssociations2(model_whz95, processedFaeces$mode1)
# testAssociations2(model_ACMTF, processedFaeces$mode1)

result_NPLS1 = testAssociationsPerComponent(NPLS_faecal)
result_NPLS2 = testAssociationsPerComponent(NPLS_milk)
result_NPLS3 = testAssociationsPerComponent(NPLS_milkMetab)

result_whz80 = testAssociationsPerComponent(model_whz80)
result_whz850 = testAssociationsPerComponent(model_whz85)
result_whz852 = testAssociationsPerComponent(model_whz852)
result_whz854 = testAssociationsPerComponent(model_whz854)
result_whz856 = testAssociationsPerComponent(model_whz856)
result_whz858 = testAssociationsPerComponent(model_whz858)
result_whz86 = testAssociationsPerComponent(model_whz86)
result_whz90 = testAssociationsPerComponent(model_whz90)
result_whz1 = testAssociationsPerComponent(model_ACMTF)

df = rbind(result_NPLS1, result_NPLS2, result_NPLS3, result_whz80, result_whz850, result_whz852, result_whz854, result_whz856, result_whz858, result_whz86, result_whz90, result_whz1)

corrected_df = signif(matrix(p.adjust(df, "BH"), dim(df)), 2)
```

```{r compare Y predictions}
cbind(model_bmi80$Yhat, model_bmi85$Yhat, model_bmi90$Yhat, model_bmi95$Yhat, Ynorm) %>%
  as_tibble() %>%
  pivot_longer(-Ynorm) %>%
  ggplot(aes(x=Ynorm,y=value,col=as.factor(name))) +
  geom_point() +
  xlab("Centered and normalized maternal ppBMI") +
  ylab("Yhat") +
  scale_color_manual(name="Model", labels=c("ACMTF-R (pi=0.80)", "ACMTF-R (pi=0.85)", "ACMTF-R (pi=0.90)", "ACMTF-R (pi=0.95)"), values=hue_pal()(5))
```

```{r comparisons of factors}
tucker_congruence <- function(x, y, ignore_sign = TRUE) {
  # x, y are numeric vectors of the same length
  if (length(x) != length(y)) stop("Vectors must have same length")
  
  numerator <- sum(x * y)
  denom_x   <- sqrt(sum(x^2))
  denom_y   <- sqrt(sum(y^2))
  
  if (denom_x < 1e-15 || denom_y < 1e-15) {
    # handle near-zero norms
    return(NA_real_)
  }
  
  phi <- numerator / (denom_x * denom_y)
  
  if (ignore_sign) {
    # Return the maximum of + or - correlation
    return(max(phi, -phi))
  } else {
    # Keep sign
    return(phi)
  }
}
```

```{r plot lambdas per model}
methods = c("pi=0.50", "pi=0.75", "pi=0.80", "pi=0.85", "pi=0.90", "pi=0.95 (1)", "pi=0.95 (2)", "pi=1 (1)", "pi=1 (2)")

cbind(model_whz50$Fac[[8]], model_whz75$Fac[[8]], model_whz80$Fac[[8]], model_whz85$Fac[[8]], model_whz90$Fac[[8]], model_whz95$Fac[[8]], model_ACMTF$Fac[[8]]) %>%
  t() %>%
  as_tibble() %>%
  mutate(model=1:9) %>%
  pivot_longer(-model) %>% 
  mutate(value=abs(value)) %>%
  ggplot(aes(x=as.factor(model),y=value,fill=as.factor(name))) +
  geom_bar(stat="identity",position=position_dodge(),col="black") +
  xlab("Model + component") +
  ylab(expression(lambda)) +
  scale_x_discrete(name="", labels=methods) +
  scale_fill_manual(name="Block", labels=c("Infant faecal microbiome", "HM microbiome", "HM metabolome"), values=hue_pal()(3)) +
  theme(legend.position="none", axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1), text=element_text(size=16))
```

```{r tucker congruence for subject loadings}
df = cbind(NPLS_faecal$Fac[[1]], NPLS_milk$Fac[[1]], NPLS_milkMetab$Fac[[1]], model_whz50$Fac[[1]], model_whz75$Fac[[1]], model_whz80$Fac[[1]], model_whz85$Fac[[1]], model_whz90$Fac[[1]], model_whz95$Fac[[1]], model_ACMTF$Fac[[1]])

congruence_matrix = matrix(NA, ncol=ncol(df), nrow=ncol(df))

for(i in 1:ncol(df)){
  for(j in 1:ncol(df)){
    congruence_matrix[i,j] = tucker_congruence(df[,i], df[,j])
  }
}

methods = c("NPLS Faecal", "NPLS Milk", "NPLS MilkMetab (1)", "NPLS MilkMetab (2)", "pi=0.50", "pi=0.75", "pi=0.80", "pi=0.85", "pi=0.90", "pi=0.95 (1)", "pi=0.95 (2)", "pi=1 (1)", "pi=1 (2)")

congruence_matrix %>%
  as_tibble() %>%
  mutate(index=1:length(methods)) %>%
  pivot_longer(-index) %>%
  mutate(name = factor(name, levels=paste0("V",1:length(methods)))) %>%
  ggplot(aes(x=as.factor(index),y=name,fill=value)) +
  geom_tile(col="black") +
  # geom_text(aes(label=round(value,2))) +
  scale_y_discrete(name="", labels=methods) +
  scale_x_discrete(name="",labels=methods) +
  scale_fill_gradient(limits=c(0,1.01)) +
  theme(legend.position="na", axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1), text=element_text(size=16))
```

```{r tucker congruence for faecal microbiome}
# NPLS_faeces_mode2 = read.csv("./Data/NPLS_milk_bmi_mode2.csv", header=FALSE) %>% as_tibble()

df = cbind(NPLS_faecal$Fac[[2]], model_whz50$Fac[[2]], model_whz75$Fac[[2]], model_whz80$Fac[[2]], model_whz85$Fac[[2]], model_whz90$Fac[[2]], model_whz95$Fac[[2]], model_ACMTF$Fac[[2]])

congruence_matrix = matrix(NA, ncol=ncol(df), nrow=ncol(df))

for(i in 1:ncol(df)){
  for(j in 1:ncol(df)){
    congruence_matrix[i,j] = tucker_congruence(df[,i], df[,j])
  }
}

methods = c("NPLS Faecal", "pi=0.50", "pi=0.75", "pi=0.80", "pi=0.85", "pi=0.90", "pi=0.95 (1)", "pi=0.95 (2)", "pi=1 (1)", "pi=1 (2)")

congruence_matrix %>%
  as_tibble() %>%
  mutate(index=1:length(methods)) %>%
  pivot_longer(-index) %>%
  mutate(name = factor(name, levels=paste0("V",1:length(methods)))) %>%
  ggplot(aes(x=as.factor(index),y=name,fill=value)) +
  geom_tile(col="black") +
  # geom_text(aes(label=round(value,2))) +
  scale_y_discrete(name="", labels=methods) +
  scale_x_discrete(name="",labels=methods) +
  scale_fill_gradient(limits=c(0,1.01)) +
  theme(legend.position="na", axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1), text=element_text(size=16))
```

```{r tucker congruence for HM microbiome}
df = cbind(NPLS_milk$Fac[[2]], model_whz50$Fac[[4]], model_whz75$Fac[[4]], model_whz80$Fac[[4]], model_whz85$Fac[[4]], model_whz90$Fac[[4]], model_whz95$Fac[[4]], model_ACMTF$Fac[[4]])

congruence_matrix = matrix(NA, ncol=ncol(df), nrow=ncol(df))

for(i in 1:ncol(df)){
  for(j in 1:ncol(df)){
    congruence_matrix[i,j] = tucker_congruence(df[,i], df[,j])
  }
}

methods = c("NPLS Milk", "pi=0.50", "pi=0.75", "pi=0.80", "pi=0.85", "pi=0.90", "pi=0.95 (1)", "pi=0.95 (2)", "pi=1 (1)", "pi=1 (2)")

congruence_matrix %>%
  as_tibble() %>%
  mutate(index=1:length(methods)) %>%
  pivot_longer(-index) %>%
  mutate(name = factor(name, levels=paste0("V",1:length(methods)))) %>%
  ggplot(aes(x=as.factor(index),y=name,fill=value)) +
  geom_tile(col="black") +
  # geom_text(aes(label=round(value,2))) +
  scale_y_discrete(name="", labels=methods) +
  scale_x_discrete(name="",labels=methods) +
  scale_fill_gradient(limits=c(0,1.01)) +
  theme(legend.position="na", axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1), text=element_text(size=16))
```

```{r tucker congruence for HM metabolome}
df = cbind(NPLS_milkMetab$Fac[[2]], model_whz50$Fac[[6]], model_whz75$Fac[[6]], model_whz80$Fac[[6]], model_whz85$Fac[[6]], model_whz90$Fac[[6]], model_whz95$Fac[[6]], model_ACMTF$Fac[[6]])

congruence_matrix = matrix(NA, ncol=ncol(df), nrow=ncol(df))

for(i in 1:ncol(df)){
  for(j in 1:ncol(df)){
    congruence_matrix[i,j] = tucker_congruence(df[,i], df[,j])
  }
}

methods = c("NPLS MilkMetab (1)", "NPLS MilkMetab (2)", "pi=0.50", "pi=0.75", "pi=0.80", "pi=0.85", "pi=0.90", "pi=0.95 (1)", "pi=0.95 (2)", "pi=1 (1)", "pi=1 (2)")
congruence_matrix %>%
  as_tibble() %>%
  mutate(index=1:length(methods)) %>%
  pivot_longer(-index) %>%
  mutate(name = factor(name, levels=paste0("V",1:length(methods)))) %>%
  ggplot(aes(x=as.factor(index),y=name,fill=value)) +
  geom_tile(col="black") +
  # geom_text(aes(label=round(value,2))) +
  scale_y_discrete(name="", labels=methods) +
  scale_x_discrete(name="",labels=methods) +
  scale_fill_gradient(limits=c(0,1.01)) +
  theme(legend.position="na", axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1), text=element_text(size=16))
```
```{r further examine the subject loadings}
comparisons = list(c("0","1"))

# NPLS
processedFaeces$mode1 %>%
  mutate(Component_1=NPLS_faecal$Fac[[1]][,1]) %>%
  ggplot(aes(x=as.factor(Secretor),y=Component_1,fill=as.factor(Secretor))) +
  geom_boxplot() +
  stat_compare_means(comparisons=comparisons,label="p.signif") +
  xlab("Secretor status") +
  scale_x_discrete(labels=c("Se-", "Se+")) +
  ylab("Component 1") +
  scale_fill_manual(name="Secretor status", labels=c("Se-", "Se+"), values=hue_pal()(2)) +
  ggtitle("NPLS faecal")

processedMilk$mode1 %>%
  mutate(Component_1=NPLS_milk$Fac[[1]][,1]) %>%
  ggplot(aes(x=as.factor(Secretor),y=Component_1,fill=as.factor(Secretor))) +
  geom_boxplot() +
  stat_compare_means(comparisons=comparisons,label="p.signif") +
  xlab("Secretor status") +
  scale_x_discrete(labels=c("Se-", "Se+")) +
  ylab("Component 1") +
  scale_fill_manual(name="Secretor status", labels=c("Se-", "Se+"), values=hue_pal()(2)) +
  ggtitle("NPLS milk")

processedMilkMetab$mode1 %>%
  mutate(Component_1=NPLS_milkMetab$Fac[[1]][,1]) %>%
  ggplot(aes(x=as.factor(Secretor),y=Component_1,fill=as.factor(Secretor))) +
  geom_boxplot() +
  stat_compare_means(comparisons=comparisons,label="p.signif") +
  xlab("Secretor status") +
  scale_x_discrete(labels=c("Se-", "Se+")) +
  ylab("Component 1") +
  scale_fill_manual(name="Secretor status", labels=c("Se-", "Se+"), values=hue_pal()(2)) +
  ggtitle("NPLS milkMetab")

# ACMTF-R
processedFaeces$mode1 %>%
  mutate(Component_1=model_bmi80$Fac[[1]][,1]) %>%
  ggplot(aes(x=as.factor(Secretor),y=Component_1,fill=as.factor(Secretor))) +
  geom_boxplot() +
  stat_compare_means(comparisons=comparisons,label="p.signif") +
  xlab("Secretor status") +
  scale_x_discrete(labels=c("Se-", "Se+")) +
  ylab("Component 1") +
  scale_fill_manual(name="Secretor status", labels=c("Se-", "Se+"), values=hue_pal()(2)) +
  ggtitle("ACMTF-R pi=0.80 (1)")

processedFaeces$mode1 %>%
  mutate(Component_2=model_bmi80$Fac[[1]][,2]) %>%
  ggplot(aes(x=as.factor(Secretor),y=Component_2,fill=as.factor(Secretor))) +
  geom_boxplot() +
  stat_compare_means(comparisons=comparisons,label="p.signif") +
  xlab("Secretor status") +
  scale_x_discrete(labels=c("Se-", "Se+")) +
  ylab("Component 2") +
  scale_fill_manual(name="Secretor status", labels=c("Se-", "Se+"), values=hue_pal()(2)) +
  ggtitle("ACMTF-R pi=0.80 (2)")

# ACMTF
processedFaeces$mode1 %>%
  mutate(Component_1=model_ACMTF$Fac[[1]][,1]) %>%
  ggplot(aes(x=as.factor(Secretor),y=Component_1,fill=as.factor(Secretor))) +
  geom_boxplot() +
  stat_compare_means(comparisons=comparisons,label="p.signif") +
  xlab("Secretor status") +
  scale_x_discrete(labels=c("Se-", "Se+")) +
  ylab("Component 1") +
  scale_fill_manual(name="Secretor status", labels=c("Se-", "Se+"), values=hue_pal()(2)) +
  ggtitle("ACMTF")

adonis2(NPLS_faecal$Fac[[1]] ~ BMI.group + C.section + Secretor + Lewis, data=processedFaeces$mode1, method="euclidean", by="margin", permutations=9999, na.action=na.omit)
adonis2(NPLS_milk$Fac[[1]] ~ BMI.group + C.section + Secretor + Lewis, data=processedMilk$mode1, method="euclidean", by="margin", permutations=9999, na.action=na.omit)
adonis2(NPLS_milkMetab$Fac[[1]] ~ BMI.group + C.section + Secretor + Lewis, data=processedMilkMetab$mode1, method="euclidean", by="margin", permutations=9999, na.action=na.omit)
adonis2(model_bmi80$Fac[[1]] ~ BMI.group + C.section + Secretor + Lewis, data=processedFaeces$mode1, method="euclidean", by="margin", permutations=9999, na.action=na.omit)
adonis2(model_bmi85$Fac[[1]] ~ BMI.group + C.section + Secretor + Lewis, data=processedFaeces$mode1, method="euclidean", by="margin", permutations=9999, na.action=na.omit)
adonis2(model_bmi90$Fac[[1]] ~ BMI.group + C.section + Secretor + Lewis, data=processedFaeces$mode1, method="euclidean", by="margin", permutations=9999, na.action=na.omit)
adonis2(model_bmi95$Fac[[1]] ~ BMI.group + C.section + Secretor + Lewis, data=processedFaeces$mode1, method="euclidean", by="margin", permutations=9999, na.action=na.omit)
adonis2(model_ACMTF$Fac[[1]] ~ BMI.group + C.section + Secretor + Lewis, data=processedFaeces$mode1, method="euclidean", by="margin", permutations=9999, na.action=na.omit)
```

```{r experimental plot}
df = cbind(NPLS_milkMetab$Fac[[2]], model_whz80$Fac[[6]], model_whz85$Fac[[6]], model_whz852$Fac[[6]], -1*model_whz854$Fac[[6]], -1*model_whz856$Fac[[6]], -1*model_whz858$Fac[[6]], model_whz86$Fac[[6]], model_whz90$Fac[[6]], model_ACMTF$Fac[[6]])
df %>% as_tibble() %>% pivot_longer(-V1) %>% ggplot(aes(x=V1,y=value,col=as.factor(name))) + geom_point() + geom_abline()
```

```{r further examine the loadings for metabolomics}
# NPLS
temp = processedMilkMetab$mode2 %>%
  mutate(Component_1=NPLS_milkMetab$Fac[[2]][,1]) %>%
  arrange(desc(Component_1)) %>%
  mutate(index=1:nrow(.))

a = temp %>%
  ggplot(aes(x=Component_1,y=as.factor(index),fill=as.factor(Class))) +
  geom_bar(stat="identity") +
  scale_y_discrete(labels=temp$Metabolite) +
  ggtitle("NPLS")



# ACMTF-R pi=0.85
temp = processedMilkMetab$mode2 %>%
  mutate(Component_1=model_whz85$Fac[[6]][,1]) %>%
  arrange(desc(Component_1)) %>%
  mutate(index=1:nrow(.))

b = temp %>%
  ggplot(aes(x=Component_1,y=as.factor(index),fill=as.factor(Class))) +
  geom_bar(stat="identity") +
  scale_y_discrete(labels=temp$Metabolite) +
  ggtitle("ACMTF-R pi=0.80")

# ACMTF-R pi=0.84
temp = processedMilkMetab$mode2 %>%
  mutate(Component_1=-1*model_whz854$Fac[[6]][,1]) %>%
  arrange(desc(Component_1)) %>%
  mutate(index=1:nrow(.))

c = temp %>%
  ggplot(aes(x=Component_1,y=as.factor(index),fill=as.factor(Class))) +
  geom_bar(stat="identity") +
  scale_y_discrete(labels=temp$Metabolite) +
  ggtitle("ACMTF-R pi=0.854")

# ACMTF-R pi=0.858
temp = processedMilkMetab$mode2 %>%
  mutate(Component_1=-1*model_whz858$Fac[[6]][,1]) %>%
  arrange(desc(Component_1)) %>%
  mutate(index=1:nrow(.))

d = temp %>%
  ggplot(aes(x=Component_1,y=as.factor(index),fill=as.factor(Class))) +
  geom_bar(stat="identity") +
  scale_y_discrete(labels=temp$Metabolite) +
  ggtitle("ACMTF-R pi=0.858")

# ACMTF
temp = processedMilkMetab$mode2 %>%
  mutate(Component_1=-1*model_ACMTF$Fac[[6]][,1]) %>%
  arrange(desc(Component_1)) %>%
  mutate(index=1:nrow(.))

e = temp %>%
  ggplot(aes(x=Component_1,y=as.factor(index),fill=as.factor(Class))) +
  geom_bar(stat="identity") +
  scale_y_discrete(labels=temp$Metabolite) +
  ggtitle("ACMTF")

ggarrange(a,b,c,d,e, common.legend=TRUE, nrow=1)
```

```{r another experimental plot}
df = cbind(NPLS_milkMetab$Fac[[2]], model_whz80$Fac[[6]], model_whz85$Fac[[6]], model_whz852$Fac[[6]], -1*model_whz854$Fac[[6]], -1*model_whz856$Fac[[6]], -1*model_whz858$Fac[[6]], model_whz86$Fac[[6]], -1*model_whz90$Fac[[6]], -1*model_ACMTF$Fac[[6]])
colnames(df) = c("NPLS Metab", "pi=0.80", "pi=0.85", "pi=0.852", "pi=0.854", "pi=0.856", "pi=0.858", "pi=0.86", "pi=0.90", "pi=1")

df %>% as_tibble() %>% mutate(Metabolite = processedMilkMetab$mode2$Metabolite, Class = processedMilkMetab$mode2$Class) %>% filter(Class == "Oligosaccharides") %>% select(-Class) %>% pivot_longer(-Metabolite) %>% ggplot(aes(x=as.factor(name),y=value,col=as.factor(Metabolite),group=as.factor(Metabolite))) + geom_line() + geom_point()
```

```{r plot the associations}
threshold = 0

# HM metabolome
df = Jakobsen2025$milk_metabolites %>% mutate(Component_1 = model_bmi95$Fac[[6]][,1])
colnames(df) = c("Metabolite", "CAS", "Component_1")

metaboliteCategories = read.csv("./Data/Metabolite_categories.txt", sep="\t") %>% as_tibble()
metaboliteCategories = metaboliteCategories %>% mutate(Metabolite = vctrs::vec_as_names(Metabolite, repair="universal_quiet"))
df = df %>% left_join(metaboliteCategories)

## Add mismatches by hand
df[df$Metabolite == "X2.Aminobutyrate","Class"] = "Amino acids and derivatives"
df[df$Metabolite == "X2.Fucosyllactose","Class"] = "Oligosaccharides"
df[df$Metabolite == "X2.Hydroxybutyrate","Class"] = "Amino acids and derivatives"
df[df$Metabolite == "X2.Oxoglutarate","Class"] = "Energy related"
df[df$Metabolite == "X3.Fucosyllactose","Class"] = "Oligosaccharides"
df[df$Metabolite == "X3SL.partial","Class"] = "Oligosaccharides"
df[df$Metabolite == "X6SL.partial","Class"] = "Oligosaccharides"
df[df$Metabolite == "Methionine","Class"] = "Amino acids and derivatives"

temp = df %>% arrange(Component_1) %>% filter(abs(Component_1) >= threshold) %>% mutate(index=1:nrow(.))
a = temp %>% ggplot(aes(x=Component_1,y=as.factor(index),fill=as.factor(Class))) + geom_bar_pattern(stat="identity", pattern_angle=45, pattern_density=0.05, pattern_spacing=0.02, pattern_key_scale_factor=0.6, col="black", pattern_fill="black") + ylab("Metabolite") + xlab("Component 1") + scale_y_discrete(labels=temp$Metabolite) + scale_fill_manual(name="Metabolite class", labels=c("Amino acids and derivatives", "Energy related", "Fatty acids and derivatives", "Food derived", "Microbially derived", "Oligosaccharides", "Pharmaceuticals", "Sugars"), values=phylum_cols) + theme(legend.position="none") + ylab("")


####

# Milk microbiome
df = Jakobsen2025$milk_microbiome_taxonomy %>% mutate(Component_1 = model_bmi95$Fac[[4]][,1]) %>% select(V3,V7,V8,Component_1)
colnames(df) = c("Phylum", "Genus", "Species", "Component_1")
temp = df %>% 
  arrange(Component_1) %>% 
  mutate(Genus = gsub("g__", "", Genus), Species = gsub("s__", "", Species)) %>% 
  mutate(Species = str_split_fixed(Species, "_", 2)[,2]) %>% 
  mutate(dplyr::across(.cols = Species, .fns = ~ dplyr::if_else(stringr::str_detect(.x, "sp."), "", .x))) %>%
  mutate(dplyr::across(.cols = Species, .fns = ~ dplyr::if_else(stringr::str_detect(.x, "organism"), "", .x))) %>%
  mutate(dplyr::across(.cols = Species, .fns = ~ dplyr::if_else(stringr::str_detect(.x, "bacterium"), "", .x)))

temp[temp$Species == "", "Species"] = "sp."
temp$name = paste0(temp$Genus, " ", temp$Species)
temp[temp$name == " sp.", "name"] = "Unclassified"
temp = temp %>% filter(abs(Component_1) >= threshold) %>% arrange(Component_1) %>% mutate(index=1:nrow(.))

b=temp %>% ggplot(aes(x=Component_1,y=as.factor(index),fill=as.factor(Phylum))) + geom_bar(stat="identity", col="black") + ylab("zOTU") + xlab("Component 1") + scale_y_discrete(labels=temp$name) + scale_fill_manual(name="Phylum", labels=c("Actinobacteriota", "Bacteroidota", "Firmicutes", "Proteobacteria"), values=phylum_cols[-3]) + theme(legend.position="none", axis.text.y = element_text(face="italic")) + ylab("")

# Faecal microbiome
df = Jakobsen2025$faecal_microbiome_taxonomy %>% mutate(Component_1 = model_bmi95$Fac[[2]][,1]) %>% select(V3,V7,V8,Component_1)
colnames(df) = c("Phylum", "Genus", "Species", "Component_1")
temp = df %>% 
  arrange(Component_1) %>% 
  mutate(Genus = gsub("g__", "", Genus), Species = gsub("s__", "", Species)) %>% 
  mutate(Species = str_split_fixed(Species, "_", 2)[,2]) %>% 
  mutate(dplyr::across(.cols = Species, .fns = ~ dplyr::if_else(stringr::str_detect(.x, "sp."), "", .x))) %>%
  mutate(dplyr::across(.cols = Species, .fns = ~ dplyr::if_else(stringr::str_detect(.x, "organism"), "", .x))) %>%
  mutate(dplyr::across(.cols = Species, .fns = ~ dplyr::if_else(stringr::str_detect(.x, "bacterium"), "", .x)))

temp[temp$Species == "", "Species"] = "sp."
temp$name = paste0(temp$Genus, " ", temp$Species)
temp[temp$name == " sp.", "name"] = "Unclassified"
temp = temp %>% filter(abs(Component_1) >= threshold) %>% arrange(Component_1) %>% mutate(index=1:nrow(.))

c=temp %>% ggplot(aes(x=Component_1,y=as.factor(index),fill=as.factor(Phylum))) + geom_bar(stat="identity", col="black") + ylab("zOTU") + xlab("Component 1") + scale_y_discrete(labels=temp$name) + scale_fill_manual(name="Phylum", labels=c("Actinobacteriota", "Bacteroidota", "Firmicutes", "Proteobacteria"), values=phylum_cols[-3]) + theme(legend.position="none", axis.text.y = element_text(face="italic")) + ylab("")

a
b
c
```
```{r loading loading plots}

# Faeces
NPLS_faeces = read.csv("./Data/NPLS_faeces_bmi_mode2.csv",header=FALSE) %>% as_tibble()
colnames(NPLS_faeces) = c("NPLS", "OTU", "Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species")
ACMTFR_faeces = cbind(-1*model_bmi95$Fac[[2]][,1], Jakobsen2025$faecal_microbiome_taxonomy) %>% as_tibble()
colnames(ACMTFR_faeces) = c("ACMTFR", "OTU", "Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species")

df = ACMTFR_faeces %>% left_join(NPLS_faeces)
  
temp = df %>% 
mutate(Genus = gsub("g__", "", Genus), Species = gsub("s__", "", Species)) %>% 
mutate(Species = str_split_fixed(Species, "_", 2)[,2]) %>% 
mutate(dplyr::across(.cols = Species, .fns = ~ dplyr::if_else(stringr::str_detect(.x, "sp."), "", .x))) %>%
mutate(dplyr::across(.cols = Species, .fns = ~ dplyr::if_else(stringr::str_detect(.x, "organism"), "", .x))) %>%
mutate(dplyr::across(.cols = Species, .fns = ~ dplyr::if_else(stringr::str_detect(.x, "bacterium"), "", .x)))

temp[temp$Species == "", "Species"] = "sp."
temp$name = paste0(temp$Genus, " ", temp$Species)
temp[temp$name == " sp.", "name"] = "Unclassified"

a = temp %>%
  ggplot(aes(x=ACMTFR,y=NPLS,col=as.factor(Phylum),label=name)) +
  geom_point() +
  # geom_text_repel(max.overlaps=10) +
  theme(legend.position="na")

# Milk
NPLS_milk = read.csv("./Data/NPLS_milk_bmi_mode2.csv",header=FALSE) %>% as_tibble()
colnames(NPLS_milk) = c("NPLS", "OTU", "Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species")
ACMTFR_milk = cbind(model_bmi95$Fac[[4]][,1], Jakobsen2025$milk_microbiome_taxonomy) %>% as_tibble()
colnames(ACMTFR_milk) = c("ACMTFR", "OTU", "Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species")

df = ACMTFR_milk %>% left_join(NPLS_milk)
  
temp = df %>% 
mutate(Genus = gsub("g__", "", Genus), Species = gsub("s__", "", Species)) %>% 
mutate(Species = str_split_fixed(Species, "_", 2)[,2]) %>% 
mutate(dplyr::across(.cols = Species, .fns = ~ dplyr::if_else(stringr::str_detect(.x, "sp."), "", .x))) %>%
mutate(dplyr::across(.cols = Species, .fns = ~ dplyr::if_else(stringr::str_detect(.x, "organism"), "", .x))) %>%
mutate(dplyr::across(.cols = Species, .fns = ~ dplyr::if_else(stringr::str_detect(.x, "bacterium"), "", .x)))

temp[temp$Species == "", "Species"] = "sp."
temp$name = paste0(temp$Genus, " ", temp$Species)
temp[temp$name == " sp.", "name"] = "Unclassified"

b = temp %>%
  ggplot(aes(x=ACMTFR,y=NPLS,col=as.factor(Phylum),label=name)) +
  geom_point() +
  # geom_text_repel(max.overlaps=10) +
  theme(legend.position="na")

# Milk metab
NPLS_milkMetab = read.csv("./Data/NPLS_milkMetab_bmi_mode2.csv", header=FALSE) %>% as_tibble()
colnames(NPLS_milkMetab) = c("NPLS", "Metabolite", "CAS")
ACMTFR_milkMetab = cbind(-1*model_bmi95$Fac[[6]][,1], Jakobsen2025$milk_metabolites) %>% as_tibble()
colnames(ACMTFR_milkMetab) = c("ACMTFR", "Metabolite", "CAS")

df = ACMTFR_milkMetab %>%
  left_join(NPLS_milkMetab, by="Metabolite") %>%
  left_join(metaboliteCategories)

## Add mismatches by hand
df[df$Metabolite == "X2.Aminobutyrate","Class"] = "Amino acids and derivatives"
df[df$Metabolite == "X2.Fucosyllactose","Class"] = "Oligosaccharides"
df[df$Metabolite == "X2.Hydroxybutyrate","Class"] = "Amino acids and derivatives"
df[df$Metabolite == "X2.Oxoglutarate","Class"] = "Energy related"
df[df$Metabolite == "X3.Fucosyllactose","Class"] = "Oligosaccharides"
df[df$Metabolite == "X3SL.partial","Class"] = "Oligosaccharides"
df[df$Metabolite == "X6SL.partial","Class"] = "Oligosaccharides"
df[df$Metabolite == "Methionine","Class"] = "Amino acids and derivatives"

c = df %>% 
  ggplot(aes(x=ACMTFR,y=NPLS,col=as.factor(Class),label=Metabolite)) +
  geom_point() +
  # geom_text_repel(max.overlaps=10) +
  theme(legend.position="na")

ggarrange(a,b,c, nrow=1)
```

```{r feature mode}
Xhat = CMTFtoolbox::reinflateFac(model_whz85$Fac, Z)
cor(Xhat[[1]][,1,1], newFaeces$mode1$whz.6m) # flip
cor(Xhat[[2]][,1,1], newFaeces$mode1$whz.6m) # flip
cor(Xhat[[3]][,1,1], newFaeces$mode1$whz.6m)

# Faeces
df = processedFaeces$mode2 %>% 
  mutate(Comp=-1*model_whz85$Fac[[2]][,1]) %>%
  arrange(Comp) %>%
  mutate(index=1:nrow(.), name=paste0(V7," ", V8)) %>%
  mutate(Genus = gsub("g__", "", V7), Species = gsub("s__", "", V8)) %>% 
  mutate(Species = str_split_fixed(Species, "_", 2)[,2]) %>% 
  mutate(dplyr::across(.cols = Species, .fns = ~ dplyr::if_else(stringr::str_detect(.x, "sp."), "", .x))) %>%
  mutate(dplyr::across(.cols = Species, .fns = ~ dplyr::if_else(stringr::str_detect(.x, "organism"), "", .x))) %>%
  mutate(dplyr::across(.cols = Species, .fns = ~ dplyr::if_else(stringr::str_detect(.x, "bacterium"), "", .x))) %>%
  filter(index %in% c(1:10, 84:93))

df[df$Genus == "" & df$Species == "", "Species"] = "Unknown"
df[df$Species == "", "Species"] = "sp."
df$name = paste0(df$Genus, " ", df$Species)

a=df %>%
  ggplot(aes(x=Comp,y=as.factor(index),fill=as.factor(V3))) +
  geom_bar(stat="identity", col="black") +
  scale_y_discrete(label=df$name) +
  xlab("Loading") +
  ylab("") +
  theme(text=element_text(size=16)) +
  scale_fill_manual(name="Phylum", values=hue_pal()(4), labels=c("Actinobacteriota", "Bacteroidota", "Firmicutes", "Proteobacteria"))

# Milk
df = processedMilk$mode2 %>% mutate(Comp=-1*model_whz85$Fac[[4]][,1]) %>%
  arrange(Comp) %>%
  mutate(index=1:nrow(.), name=paste0(V7," ", V8)) %>%
  mutate(Genus = gsub("g__", "", V7), Species = gsub("s__", "", V8)) %>% 
  mutate(Species = str_split_fixed(Species, "_", 2)[,2]) %>% 
  mutate(dplyr::across(.cols = Species, .fns = ~ dplyr::if_else(stringr::str_detect(.x, "sp."), "", .x))) %>%
  mutate(dplyr::across(.cols = Species, .fns = ~ dplyr::if_else(stringr::str_detect(.x, "organism"), "", .x))) %>%
  mutate(dplyr::across(.cols = Species, .fns = ~ dplyr::if_else(stringr::str_detect(.x, "bacterium"), "", .x))) %>%
  filter(index %in% c(1:10, 110:119))

df[df$Genus == "" & df$Species == "", "Species"] = "Unknown"
df[df$Species == "", "Species"] = "sp."
df$name = paste0(df$Genus, " ", df$Species)

b=df %>%
  ggplot(aes(x=Comp,y=as.factor(index),fill=as.factor(V3))) +
  geom_bar(stat="identity", col="black") +
  scale_y_discrete(label=df$name) +
  xlab("Loading") +
  ylab("") +
  theme(text=element_text(size=16)) +
  scale_fill_manual(name="Phylum", values=hue_pal()(4)[-2], labels=c("Actinobacteriota", "Firmicutes", "Proteobacteria"))

# MilkMetab
df = processedMilkMetab$mode2 %>% mutate(Comp=model_whz85$Fac[[6]][,1]) %>% arrange(Comp) %>% mutate(index=1:nrow(.), name=Metabolite) %>% filter(index %in% c(1:10, 61:70))
c=df %>%
  ggplot(aes(x=Comp,y=as.factor(index),fill=as.factor(Class),pattern=as.factor(Class))) +
  geom_bar_pattern(stat="identity",
                   colour="black",
                   pattern="stripe",
                   pattern_fill="black",
                   pattern_density=0.2,
                   pattern_spacing=0.05,
                   pattern_angle=45,
                   pattern_size=0.2) +
  scale_y_discrete(label=df$name) +
  xlab("Loading") +
  ylab("") +
  guides(fill=guide_legend(title="Class")) +
  theme(text=element_text(size=16))

```
```{r plot time mode}
d = model_whz85$Fac[[3]] %>%
  as_tibble() %>%
  mutate(time=c(30,60,90)) %>% 
  pivot_longer(-time) %>%
  ggplot(aes(x=time,y=value)) + 
  geom_line() + 
  geom_point() +
  ylim(0,1) +
  xlab("Time point [days]") +
  ylab("Loading") +
  theme(text=element_text(size=16))

e = (-1*model_whz85$Fac[[5]]) %>%
  as_tibble() %>%
  mutate(time=c(0,30,60,90)) %>% 
  pivot_longer(-time) %>%
  ggplot(aes(x=time,y=value)) + 
  geom_line() + 
  geom_point() +
  ylim(0,1) +
  xlab("Time point [days]") +
  ylab("Loading") +
  theme(text=element_text(size=16))

f = (-1*model_whz85$Fac[[7]]) %>%
  as_tibble() %>%
  mutate(time=c(0,30,60,90)) %>% 
  pivot_longer(-time) %>%
  ggplot(aes(x=time,y=value)) + 
  geom_line() + 
  geom_point() +
  ylim(0,1) +
  xlab("Time point [days]") +
  ylab("Loading") +
  theme(text=element_text(size=16))

ggarrange(a,b,c,nrow=1,ncol=3)
```
